egs/aishell/s10/chain/tdnnf_layer.py

```diff
@@ -0,0 +1,319 @@
+        kernel_size = w.size(2)
+
+        w = w.reshape(out_channels, -1)
+
+        num_rows = w.size(0)
+        num_cols = w.size(1)
+
+        need_transpose = False
+        if num_rows > num_cols:
+            w = w.t()
+            need_transpose = True
+
+        w = _constrain_orthonormal_internal(w)
+
+        if need_transpose:
+            w = w.t()
+
+        w = w.reshape(out_channels, in_channels, kernel_size)
+
+        state_dict['weight'] = w
+        self.conv.load_state_dict(state_dict)
+
+
+class PrefinalLayer(nn.Module):
+
+    def __init__(self, big_dim, small_dim):
+        super().__init__()
+        self.affine = nn.Linear(in_features=small_dim, out_features=big_dim)
+        self.batchnorm1 = nn.BatchNorm1d(num_features=big_dim)
+        self.linear = OrthonormalLinear(dim=big_dim,
+                                        bottleneck_dim=small_dim,
+                                        time_stride=0)
+        self.batchnorm2 = nn.BatchNorm1d(num_features=small_dim)
+
+    def forward(self, x):
+        # x is [N, C, T]
+        x = x.permute(0, 2, 1)
+
+        # at this point, x is [N, T, C]
+
+        x = self.affine(x)
+        x = F.relu(x)
+
+        # at this point, x is [N, T, C]
+
+        x = x.permute(0, 2, 1)
+
+        # at this point, x is [N, C, T]
+
+        x = self.batchnorm1(x)
+
+        x = self.linear(x)
+
+        x = self.batchnorm2(x)
+
+        return x
+
+    def constrain_orthonormal(self):
+        self.linear.constrain_orthonormal()
+
+
+class FactorizedTDNN(nn.Module):
+    '''
+    This class implements the following topology in kaldi:
+      tdnnf-layer name=tdnnf2 $tdnnf_opts dim=1024 bottleneck-dim=128 time-stride=1
+
+    References:
+        - http://danielpovey.com/files/2018_interspeech_tdnnf.pdf
+        - ConstrainOrthonormalInternal() from
+          https://github.com/kaldi-asr/kaldi/blob/master/src/nnet3/nnet-utils.cc#L982
+    '''
+
+    def __init__(self,
+                 dim,
+                 bottleneck_dim,
+                 time_stride,
+                 conv_stride,
+                 bypass_scale=0.66):
+        super().__init__()
+
+        assert conv_stride in [1, 3]
+        assert abs(bypass_scale) <= 1
+
+        self.bypass_scale = bypass_scale
+
+        self.conv_stride = conv_stride
+
+        # linear requires [N, C, T]
+        self.linear = OrthonormalLinear(dim=dim,
+                                        bottleneck_dim=bottleneck_dim,
+                                        time_stride=time_stride)
+
+        # affine requires [N, C, T]
+        # WARNING(fangjun): we do not use nn.Linear here
+        # since we want to use `stride`
+        self.affine = nn.Conv1d(in_channels=bottleneck_dim,
+                                out_channels=dim,
+                                kernel_size=1,
+                                stride=conv_stride)
+
+        # batchnorm requires [N, C, T]
+        self.batchnorm = nn.BatchNorm1d(num_features=dim)
```

Comment:
It would be a closer match to what Kaldi's system is doing if you were to add `affine=False` wherever you use batchnorm.

Reply:
will be addressed in the next pullrequest.